pkgs/utils/src/log.zig

```diff
@@ -31,86 +31,169 @@ pub fn compTimeLog(comptime scope: LoggerScope, activeLevel: std.log.Level, comp
         var ts_buf: [64]u8 = undefined;
         const timestamp_str = getFormattedTimestamp(&ts_buf);
 
-        nosuspend stderr.print(
+        var buf: [4096]u8 = undefined;
+        const print_str = std.fmt.bufPrint(
+            buf[0..],
             "{s} {s}" ++ fmt ++ "\n",
             .{ timestamp_str, prefix } ++ args,
         ) catch return;
+
+        // Print to stderr
+        nosuspend stderr.writeAll(print_str) catch return;
+
+        // Also write to file if provided
+        if (file) |f| {
+            nosuspend f.writeAll(print_str) catch return;
+        }
     }
 }
 
-pub fn log(scope: LoggerScope, activeLevel: std.log.Level, comptime level: std.log.Level, comptime fmt: []const u8, args: anytype) void {
+pub fn log(scope: LoggerScope, activeLevel: std.log.Level, comptime level: std.log.Level, comptime fmt: []const u8, args: anytype, file: ?std.fs.File) void {
     switch (scope) {
-        .default => return compTimeLog(.default, activeLevel, level, fmt, args),
-        .n1 => return compTimeLog(.n1, activeLevel, level, fmt, args),
-        .n2 => return compTimeLog(.n2, activeLevel, level, fmt, args),
-        .n3 => return compTimeLog(.n3, activeLevel, level, fmt, args),
+        .default => return compTimeLog(.default, activeLevel, level, fmt, args, file),
+        .n1 => return compTimeLog(.n1, activeLevel, level, fmt, args, file),
+        .n2 => return compTimeLog(.n2, activeLevel, level, fmt, args, file),
+        .n3 => return compTimeLog(.n3, activeLevel, level, fmt, args, file),
     }
 }
 
-//
 const LoggerScope = enum {
     default,
     n1,
     n2,
     n3,
 };
 
+pub const FileParams = struct {
+    filePath: ?[]const u8,
+    fileName: ?[]const u8,
```

Comment:
```suggestion
    filePath: []const u8,
    fileName: []const u8,
```
no need to make individual params optional